### PR TITLE
[1523] Get supported formats extensions dynamically

### LINKED ISSUE #1523: Dinamically get the supported format types to import
Since #1444 there is a preset list of extensions for the browser to search.
This list is hardcoded.

If a new file format is added, eg. CSV, this list isn't updated.

Solution: Get the supported formats extensions and dinamically produce this list.

### PR DESCRIPTION
**Pull request by pedro-mendonca**
## What?

Dynamically generate the supported formats file extensions for file imports.

## Why?
If a plugin adds a new file format, dynamically get all the available extensions.

## How?
Use the set Formats.

## Testing Instructions
Go to the project import dialog and click the button to upload.

Fix #1523

### CODE REVIEW

File: gp-templates/project-import.php [hunk 1]
```diff
@@ -24,14 +24,22 @@
...
 gp_title( $gp_title );
 gp_tmpl_header();
-?>
 
+$format_extensions = array();
+foreach ( GP::$formats as $format ) {
+	$format_extensions = array_merge( $format_extensions, $format->get_file_extensions() );
+}
+// Remove redundant alternate extensions that have the same ending and add leading '.'.
+$format_extensions = array_unique( preg_replace( '/^(\w*\.)*/', '.', $format_extensions ) );
```
Comment (akirk) [inline gp-templates/project-import.php]: Would this `preg_replace` be better suited to be inside `get_file_extensions()`?

**Review by akirk — COMMENTED** [review]
Could we please add a unit test for this?

**Review by akirk — CHANGES_REQUESTED** [review]
I think changes are warranted as we’re making assumptions about the file extensions inside a template file and I think it’d be better suited to put this logic inside the `GP_Format` class. We could introduce a new function like `get_acceptable_file_extension()`

File: gp-templates/project-import.php (hunk 1 above)
Comment (pedro-mendonca) [reply gp-templates/project-import.php]: Moving the `preg_replace` inside the `get_file_extensions()` will remove these from the output: `jed.json`, `resx.xml`, `ngx.json`.
I'd guess this won't be desirable and might break its use elsewhere.

Comment (akirk) [reply gp-templates/project-import.php]: Yes, you are right. I proposed in the review comment to add a new function.

**Comment by pedro-mendonca** [comment]
> We could introduce a new function like `get_acceptable_file_extension()`

Should I add it to `GP_Format`?

**Comment by akirk** [comment]
Yes, I think it'd be a good place.

**Comment by pedro-mendonca** [comment]
> Yes, I think it'd be a good place.

So you suggest a new `get_acceptable_file_extension()` that strips any first part of the extension (eg. [jed.]json). Should this method also add the "." to the extensions?

**Comment by pedro-mendonca** [comment]
With the exception of this list of extensions used in this PR where only the last part of the extension counts for the browser to look up, I don't see much of a use for a new method that uses the output of `get_file_extensions()` and strip any first part of the extension in case there are multiple parts (eg. jed.json), in the `GP_Fotmat` class.
Also, adding a "." is just for use in this browser lookup situation, considering all that needs to be done afterward, it doesn't make sense to me to add it in a GP_Format output.

A different option would be adding a function to `misc.php` where all this is done: getting all the extensions, optionally stripping the non-ending part of the extensions, removing duplicates, and adding the ".", a new `gp_get_formats_extensions()` that would output a simple array of all the formats extensions, with no duplicates.

**Comment by pedro-mendonca** [comment]
@akirk here's an example of what I have in mind:

A `function gp_get_format_extensions()` which outputs all the formats prefixed with `.` delimiter, allowing multiple extensions like `.jed.json`.

`gp_get_format_extensions()`:
```
array(
  '.xml',
  '.po',
  '.pot',
  '.mo',
  '.resx',
  '.resx.xml',
  '.strings',
  '.properties',
  '.json',
  '.jed.json',
  '.ngx.json',
)
```

The multiple extensions like `.jed.json` aren't redundant as it allows the user to select only that option, filtering out files with the single extension `.json`. Because of this, I think I should remove the `preg_replace` that strips out the multiple extensions.

Would this make sense to add to `misc.php`?

**Review by akirk — APPROVED** [review]
`misc.php` is a reasonable place since other functions also operate on `GP::$formats` there.

**Pull request merged**